### PR TITLE
TheiaMeta: Simplify metagenomics workflow

### PR DESCRIPTION
**Pull request by cimendes**
# This PR does not merge to main!

## :hammer_and_wrench: Changes Being Made

This PR implements a simplified and polished version of the TheiaMeta_Illumina_PE workflow for viral metagenomic analysis. 
The diagram of the workflow is available below:

Viral Metagenomic workflow - TheiaMeta_Illumina_PE

## :clipboard: Workflow/Task Steps

Please see the diagram above. 

This workflow takes in Illumina PE data and performs:
- taxonomical assignment of the raw read data (kraken2) 
- host removal (human), 
- quality trimming (trimmomatic)
- metagenomic genome assembly (megahit)
- IF A REFERENCE FILE IS PROVIDED: contig capture to the reference (minimap2)
- retrieval of unassembled reads

### Inputs

 This workflow takes Illumina paired-end sequencing data

### Outputs

```bash
{
    "theiameta_illumina_pe.megahit_pe_version": "${this.megahit_pe_version}",
    "theiameta_illumina_pe.num_reads_clean2": "${this.num_reads_clean2}",
    "theiameta_illumina_pe.bbduk_docker": "${this.bbduk_docker}",
    "theiameta_illumina_pe.kraken_percent_human": "${this.kraken_percent_human}",
    "theiameta_illumina_pe.kraken_version": "${this.kraken_version}",
    "theiameta_illumina_pe.read1_clean": "${this.read1_clean}",
    "theiameta_illumina_pe.read2_unmapped": "${this.read2_unmapped}",
    "theiameta_illumina_pe.num_reads_clean1": "${this.num_reads_clean1}",
    "theiameta_illumina_pe.read1_dehosted": "${this.read1_dehosted}",
    "theiameta_illumina_pe.theiameta_illumina_pe_analysis_date": "${this.theiameta_illumina_pe_analysis_date}",
    "theiameta_illumina_pe.kraken_report": "${this.kraken_report}",
    "theiameta_illumina_pe.num_reads_raw2": "${this.num_reads_raw2}",
    "theiameta_illumina_pe.read1_unmapped": "${this.read1_unmapped}",
    "theiameta_illumina_pe.read2_dehosted": "${this.read2_dehosted}",
    "theiameta_illumina_pe.num_reads_raw1": "${this.num_reads_raw1}",
    "theiameta_illumina_pe.fastq_scan_version": "${this.fastq_scan_version}",
    "theiameta_illumina_pe.read2_clean": "${this.read2_clean}",
    "theiameta_illumina_pe.num_reads_raw_pairs": "${this.num_reads_raw_pairs}",
    "theiameta_illumina_pe.trimmomatic_version": "${this.trimmomatic_version}",
    "theiameta_illumina_pe.largest_contig": "${this.largest_contig}",
    "theiameta_illumina_pe.theiameta_illumina_pe_version": "${this.theiameta_illumina_pe_version}",
    "theiameta_illumina_pe.contig_number": "${this.contig_number}",
    "theiameta_illumina_pe.assembly_fasta": "${this.assembly_fasta}",
    "theiameta_illumina_pe.num_reads_clean_pairs": "${this.num_reads_clean_pairs}"
}
```
### Locally

The local run of the TheiaMeta_Illumina_PE is successful. 

### Terra

Run with 12 monkeypox blood samples: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/79ca2243-0e85-4cca-ac95-b86d6e0de1cd

Run with 32 HAV blood samples: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/1e32eb72-8884-408f-b153-4a845a1f1ba5 (failed samples have no HAV sequence data, so they are expected)

Run with 22 DENV blood samples: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/34f0e78f-c25b-4db9-9661-d23586b6dfed (failed sample is a negative control) 

## :microscope: Quality checks

Pull Request (PR) checklist:
- [x] Include a description of what is in this pull request in this message.
- [x] The workflow/task has been tested locally and on Terra
- [ ] The CI/CD has been adjusted and tests are passing
- [x] Everything follows the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

### CODE REVIEW

**Comment by cimendes** [comment]
Tested on commit id [ff4b662](https://github.com/theiagen/public_health_bioinformatics/pull/75/commits/ff4b662153911d49656de33877c6806d75f928bc)

DENV w/ Reference: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/8d07ab03-299b-4ccc-a609-372bf2ab6f2f

HAV w/ Reference: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/e6876654-facc-4929-9aa1-bf990e59567d

**Comment by cimendes** [comment]
Documentation: https://www.notion.so/theiagen/TheiaMeta-65633b05fecc48c583cd74d000ef0364

File: tasks/quality_control/task_quast.wdl [hunk 1]
```diff
@@ -5,6 +5,7 @@ task quast {
     File assembly
     String samplename
     String docker = "quay.io/staphb/quast:5.0.2"
+    Int min_contig_len = 500
```
Comment (sage-wright) [inline tasks/quality_control/task_quast.wdl:8]: will this interfere with the other workflows that use quast?

Comment (cimendes) [reply tasks/quality_control/task_quast.wdl:8]: It shouldn't as that's already the default value implemented when you don't set that option for quast's command.

File: tasks/utilities/task_compare_assemblies.wdl [hunk 2]
```diff
@@ -21,25 +26,42 @@ task compare_assemblies {
...
+    if denovo_base_count >= consensus_base_count:
         input_file = "~{assembly_denovo}"
+        with open("NUMBER_ATCG", "w") as number_ATCG:
+          number_ATCG.write(str(denovo_base_count))
     else:
         input_file = "~{assembly_consensus}"
+        with open("NUMBER_ATCG", "w") as number_ATCG:
+          number_ATCG.write(str(consensus_base_count))
 
     with open("~{samplename}_highest.fasta", 'w') as out_f:
```
Comment (sage-wright) [inline tasks/utilities/task_compare_assemblies.wdl:46]: this section seems a bit inefficient to me. Could you not just say `~{samplename}_highest.fasta = input_file`? Then you don't have to iterate through the file.

Comment (sage-wright) [reply tasks/utilities/task_compare_assemblies.wdl:46]: Also, this task isn't being called in the workflow. Maybe I'm going over this too soon and you're still working on it though 😂

Comment (cimendes) [reply tasks/utilities/task_compare_assemblies.wdl:46]: You're right. :) I wrote a lot of tasks that in the end I didn't end up using, like the bcftools one to call consensus sequences. I'm actually not sure if I should remove everything or leave them be in the case they are useful in the future

Comment (cimendes) [reply tasks/utilities/task_compare_assemblies.wdl:46]: I'll be removing the unused tasks that were included in this PR :)

**Comment by sage-wright** [comment]
Ran a sanity check here: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/ce41828f-ef04-4fb8-80e9-84b0dc71d5f3

Looks great, Inês!

**Pull request merged**